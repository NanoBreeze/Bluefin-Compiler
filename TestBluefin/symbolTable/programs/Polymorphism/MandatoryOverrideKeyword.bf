/*
The 'override' keyword must be applied to a method if its base class (or recursively deep base class) has a virtual method with 
identical name and signature (param types and return type). 
Even if in between grandparent and current class, there's a variable with same name as function, as long as base class has the 
virtual method, we must apply override keyword (eg, look at variable 'h' in Der)
*/

struct Base {
	virtual void f(int a) {}
	virtual float g() {}
	virtual int h() {}

	void x() {}
};

struct Der extends Base {
	void f(int b) {}		// Must apply 'override' keyword
	int h;					// note this is the same name as Base::h(), ok
};

struct DerDer extends Der {
	void f(int b) {}		// Must apply 'override' keyword b/c grandparent has virtual method
	float g() {}			// Must apply 'override' b/c identical signature to grandparent's even though parent's isn't
	int h() {}				// Must apply 'override' 

	void x() {}				// OK, b/c Base::x() isn't virtual

};