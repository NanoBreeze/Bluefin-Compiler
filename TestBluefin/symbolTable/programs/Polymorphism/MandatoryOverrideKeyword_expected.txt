Missing 'override' specifier for method 'f'
Missing 'override' specifier for method 'f'
Missing 'override' specifier for method 'g'
Missing 'override' specifier for method 'h'
